### PR TITLE
Change `norm_factor` into `softmax_scale` and add kwarg into `DotProductAttention `

### PR DESCRIPTION
**Pull request by BoxiangW**
# Description

Add arg option of `norm_factor` into DotProductAttention for MLA support

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by BoxiangW** [comment]
Hi @cyanguwa , could you help take a look at this PR?

**Review by ksivaman — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/attention.py
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:3410]: I don't think this is a good name since it doesn't involve norms. `softmax_scale` (from [`flash_attn.flash_attn_qkvpacked_func`](https://github.com/Dao-AILab/flash-attention/blob/320fb59487658f033f56711efd3d61b7c7a6f8f3/flash_attn/flash_attn_interface.py#L642) or `scale` (from [`torch.nn.functional.scaled_dot_product_attention`](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch-nn-functional-scaled-dot-product-attention)) is more descriptive:
```suggestion
        softmax_scale: Optional[float] = None,
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/attention.py
Comment (BoxiangW) [reply transformer_engine/pytorch/attention.py]: I have changed the name of this kwarg

**Review by timmoon10 — APPROVED** [review]
LGTM. Can you sign your last commit?

**Comment by BoxiangW** [comment]
Hi @timmoon10 , I have signed off my commits. Seems like CIs are waiting for approval right now.

**Comment by timmoon10** [comment]
Ah, actually I was incorrect. The `softmax_scale` we use here is the reciprocal of the `softmax_scale` used in Flash Attention, so `norm_factor` is a better name. Better yet, we would want to refactor our implementation so we have the same scaling factor convention as Flash Attention and PyTorch.

**Comment by BoxiangW** [comment]
> Ah, actually I was incorrect. The `softmax_scale` we use here is the reciprocal of the `softmax_scale` used in Flash Attention, so `norm_factor` is a better name. Better yet, we would want to refactor our implementation so we have the same scaling factor convention as Flash Attention and PyTorch.

Yes, I found this part under `FlashAttention` where softmax_scale is 1.0/norm_factor. So I am going to change the kwarg name back to `norm_factor`
```python
            with self.attention_dropout_ctx():
                output = attn_forward_func_with_cp(
                    self.training, query_layer, key_layer, value_layer,
                    cu_seqlens_q, cu_seqlens_kv, max_seqlen_q, max_seqlen_kv,
                    self.attention_dropout if self.training else 0.0,
                    cp_group, cp_global_ranks, cp_stream,
                    softmax_scale=1.0/self.norm_factor,
                    qkv_format="bshd" if qkv_format=="sbhd" else qkv_format,
                    attn_mask_type=attn_mask_type,
                    deterministic=self.deterministic
                )
```

File: transformer_engine/pytorch/attention.py
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:3386]: The quickest way to keep the `softmax_scale` kwarg consistent with Flash Attention:
```suggestion
        if softmax_scale is None:
            softmax_scale = math.sqrt(kv_channels)
        norm_factor = 1 / softmax_scale
```

Comment (BoxiangW) [reply transformer_engine/pytorch/attention.py:3386]: I am ok with this implementation, not so sure whether this is a wanted implementation for TE's code readability though. What do you think

Comment (timmoon10) [reply transformer_engine/pytorch/attention.py:3386]: My thinking is that it's more important to get the API right. If we add something non-standard like `norm_factor`, then it'll confuse users and be painful to remove in the future. In the future, I think it would be nice to refactor to use `softmax_scale` instead of `norm_factor` in all of the attention backends.

What do you think @cyanguwa?

File: transformer_engine/pytorch/attention.py
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1733]:
```suggestion
            scale /= self.layer_number
```

File: transformer_engine/pytorch/attention.py
Comment (BoxiangW) [reply transformer_engine/pytorch/attention.py]: Yes, you are right. I will change this typo

**Review by cyanguwa — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/attention.py
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py]: I think keeping things as `softmax_scale = 1/sqrt(head_dim)` is good!

**Comment by BoxiangW** [comment]
Seems like there's a failing test, but after I look at the details, it says all successful.

**Pull request merged**
